// teeworlds-econ/lib/handlers/index.js
var handleCaptureMessage, handleChatMessage, handleEnterMessage, handleFlagGrabMessage, handleFlagReturnMessage, handleKillMessage, handleLeaveMessage, handleNetBanMessage, handlePickupMessage;

handleEnterMessage = require('./handle-enter-message');

handleLeaveMessage = require('./handle-leave-message');

handlePickupMessage = require('./handle-pickup-message');

handleChatMessage = require('./handle-chat-message');

handleKillMessage = require('./handle-kill-message');

handleFlagGrabMessage = require('./handle-flag-grab-message');

handleFlagReturnMessage = require('./handle-flag-return-message');

handleCaptureMessage = require('./handle-capture-message');

handleNetBanMessage = require('./handle-net-ban-message');

module.exports = {
  handleEnterMessage: handleEnterMessage,
  handleLeaveMessage: handleLeaveMessage,
  handlePickupMessage: handlePickupMessage,
  handleChatMessage: handleChatMessage,
  handleKillMessage: handleKillMessage,
  handleFlagGrabMessage: handleFlagGrabMessage,
  handleFlagReturnMessage: handleFlagReturnMessage,
  handleCaptureMessage: handleCaptureMessage,
  handleNetBanMessage: handleNetBanMessage
};
